### PR TITLE
Allow `--local-crds` flag to receive multiple values

### LINKED ISSUE #100: Allow `--local-crds` flag to receive multiple values
### What would you like to be added?

The `--local-crds` flag currently supports a single value, which restricts loading the CRDs to a single directory.

As a result, invoking kubectl-validate like so will only consider CRDs from `path/to/second/dir`:
```
kubectl-validate --local-crds path/to/first/dir --local-crds path/to/second/dir path/to/resources/*.yaml
```

We could allow this flag to receive multiple values to allow collecting CRDs from multiples directories.

### Why is this needed?

The current behavior requires users to have all their CRDs in a single directory, which might require an extra effort if it is not how they initially store them.

The current behavior is also a bit counter-intuitive as the flag can be specified multiple times but will only use the value that was provided last.

### PR DESCRIPTION
**Pull request by doom**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

This PR allows the `--local-crds` flag to receive multiple values.

#### Which issue(s) this PR fixes:

Fixes #100

#### Special notes for your reviewer:

N/A

#### Does this PR introduce a user-facing change?

```release-note
Updated the `--local-crds` flag to allow receiving multiple values
```

### CODE REVIEW

File: pkg/cmd/validate.go [hunk 1]
```diff
@@ -59,7 +59,7 @@ type commandFlags struct {
 	kubeConfigOverrides clientcmd.ConfigOverrides
 	version             string
 	localSchemasDir     string
-	localCRDsDir        string
+	localCRDsDir        *[]string
```
Comment (apelisse) [inline pkg/cmd/validate.go:62]: Why pointer here? That looks surprising

File: pkg/cmd/validate.go [hunk 2]
```diff
@@ -79,7 +79,7 @@ func NewRootCommand() *cobra.Command {
 	}
 	res.Flags().StringVarP(&invoked.version, "version", "", invoked.version, "Kubernetes version to validate native resources against. Required if not connected directly to cluster")
 	res.Flags().StringVarP(&invoked.localSchemasDir, "local-schemas", "", "", "--local-schemas=./path/to/schemas/dir. Path to a directory with format: /apis/<group>/<version>.json for each group-version's schema.")
-	res.Flags().StringVarP(&invoked.localCRDsDir, "local-crds", "", "", "--local-crds=./path/to/crds/dir. Path to a directory containing .yaml or .yml files for CRD definitions.")
+	invoked.localCRDsDir = res.Flags().StringSliceP("local-crds", "", []string{}, "--local-crds=./path/to/crds/dir. Path to a directory containing .yaml or .yml files for CRD definitions.")
```
Comment (apelisse) [inline pkg/cmd/validate.go:82]:
```suggestion
	invoked.localCRDsDir = res.Flags().StringSliceP("local-crds", "", []string{}, "--local-crds=./path/to/crds/dir. Paths to directories containing .yaml or .yml files for CRD definitions.")
```

File: pkg/openapiclient/local_crds.go [hunk 3]
```diff
@@ -33,55 +33,60 @@ var metadataSchemas map[string]*spec.Schema = func() map[string]*spec.Schema {
...
 	var documents []utils.Document
-	for _, f := range files {
-		path := filepath.Join(k.dir, f.Name())
-		if f.IsDir() {
-			continue
+
+	for _, dir := range k.dirs {
+		files, err := utils.ReadDir(k.fs, dir)
+		if err != nil {
+			return nil, fmt.Errorf("error listing %s: %w", dir, err)
```
Comment (apelisse) [inline pkg/openapiclient/local_crds.go:58]: Do we want to stop if any of the directory fails? I think we do
But do we want to report all errors if multiple directories fail? I suspect so too, thoughts?

File: pkg/openapiclient/local_crds.go [hunk 4]
```diff
@@ -33,55 +33,60 @@ var metadataSchemas map[string]*spec.Schema = func() map[string]*spec.Schema {
...
-			if err != nil {
-				return nil, fmt.Errorf("failed to read %s: %w", path, err)
-			}
-			yamlDocs, err := utils.SplitYamlDocuments(fileBytes)
-			if err != nil {
-				return nil, fmt.Errorf("failed to read %s: %w", path, err)
+
+		for _, f := range files {
+			path := filepath.Join(dir, f.Name())
+			if f.IsDir() {
```
Comment (apelisse) [inline pkg/openapiclient/local_crds.go:63]: I didn't realize this wasn't recursive, that feels weird. But it has nothing to do with this PR so forget it for now.

File: pkg/openapiclient/local_crds.go (hunk 3 above)
Comment (doom) [reply pkg/openapiclient/local_crds.go:58]: I agree for a better user experience we likely should accumulate and report all errors at once.
However it seems the current strategy for this function is to fail as soon as any error is encountered (e.g. first file that cannot be read). Perhaps we could revisit it outside of this PR ?

**Comment by kannon92** [comment]
/ok-to-test

Not sure why the CI is stuck..

**Comment by doom** [comment]
> /ok-to-test
> 
> Not sure why the CI is stuck..

Hey @kannon92 ! Not very familiar with the process, but from my side the workflow is still marked as awaiting approval from a maintainer (and links to [these docs](https://docs.github.com/en/actions/managing-workflow-runs/approving-workflow-runs-from-public-forks#approving-workflow-runs-on-a-pull-request-from-a-public-fork)). Would you or @apelisse be able to do that ?

**Comment by alexzielenski** [comment]
I marked it as ok to test

**Comment by doom** [comment]
Hey @alexzielenski, thank you ! I have rebased the branch on top of `main` to fix `go.mod` diffs, could you please re-approve the workflow run ?

**Comment by alexzielenski** [comment]
/approve

Thanks for this! Will defer lgtm for a bit in case @apelisse has anything to add

File: pkg/openapiclient/local_crds.go (hunk 3 above)
Comment (alexzielenski) [reply pkg/openapiclient/local_crds.go:58]: sgtm

**Comment by monotek** [comment]
I ran into an issue recently, where in my CRDs directory, where i had to copy all my CRDs, a non CRD yaml file existed, which caused kubectl-validate to fail, like below.

```bash
kubectl-validate namespaces/visual-regression-tracker --local-crds ./crds
Error: no kind "ClusterPolicy" is registered for version "kyverno.io/v1" in scheme "pkg/runtime/scheme.go:100"
```

So if multiple directories would be possible for --local-crds i guess there should be some check if the files in the directories are valid CRDs. 
I guess checking for "^kind: CustomResourceDefinition$" in the file is is enough?
All other files should be ignored.

**Comment by alexzielenski** [comment]
> So if multiple directories would be possible for --local-crds imho there should be some check if the files in the directories are valid CRDs.
I guess checking for "kind: CustomResourceDefinition" in the file is is enough?
All other files should be ignored.

That is valid, I agree we should ignore non-crds in those directories. Could you file a separate issue to track this?

**Pull request merged**